### ISSUE TITLE
Tokenizer Training

### ISSUE THREAD

**Issue by yukiarimo** [QwenLM/Qwen3-TTS#214]
### Description

Can you please provide full code how to fully train (or fine-tune) Qwen’s TTS Tokenizer.

Yes, I know it is based on Mimi. Mimi doesn’t provided training code. I need a training exactly for THIS model!

Thanks!

### Reproduction

.

### Logs

```shell

```

### Environment Information

Colab

### Known Issue

- [x] The issue hasn't been already addressed in Documentation, Issues, and Discussions.

**Comment by takuma104** [comment]
I was also waiting for the official training code, but since it hasn't been released no matter how long I wait, I made my own. Feel free to use it if you like. This is a script that performs fine-tuning only on the decoder (so codec compatibility is preserved). It also comes with an added bonus: you can tweak the upsampler settings for 48kHz output.
https://github.com/takuma104/Qwen3-TTS-Tokenizer-12Hz-Trainer

**Comment by yukiarimo** [comment]
Wooooooah!!!! Thank you very much! I’ll definitely look into it. Few quick questions:

1. Can it be trained with 50 hours of data? Or what is the minimum requirement (for both fine-tuning and from scratch)?
2. Have YOU trained running the training loop on something, yet?
3. What is that 48 kHz upsampling block (haven’t looked into code yet)? Is it a usual CNN upsampling or external stretching? Aka, how does it work?
4. Is it possible to train it on music? (Just for fun experiment)

**Comment by takuma104** [comment]
> Can it be trained with 50 hours of data? Or what is the minimum requirement (for both fine-tuning and from scratch)? 

Regarding from-scratch training, assuming it's comparable to other neural codecs, I believe the minimum requirement is on the order of tens of thousands of hours. It seems around 500 hours would be needed for 48kHz fine-tuning. However, since I only trained for one epoch, how it performs with multiple epochs remains unknown. For example, if fine-tuning at 24kHz, dozens of hours might suffice even with multiple epochs. 

> Have YOU trained running the training loop on something, yet? 

The model weights [already released](https://huggingface.co/takuma104/Qwen3-TTS-Tokenizer-12Hz-48kHz) are from the previous codebase. To verify the current codebase, which includes some minor bug fixes, I am currently running '[Run7](https://wandb.ai/takuma104/Qwen3-TTS-Tokenizer-12Hz-Trainer/runs/nwhaeaxq?nw=nwusertakuma104)'. The loss should probably stabilize within the next 2-3 days, and I plan to release that then. 

> What is that 48 kHz upsampling block (haven’t looked into code yet)? Is it a usual CNN upsampling or external stretching? Aka, how does it work? 

For a simple explanation, please refer [here](https://github.com/takuma104/Qwen3-TTS-Tokenizer-12Hz-Trainer?tab=readme-ov-file#how-it-works). In the training code, [this line](https://github.com/takuma104/Qwen3-TTS-Tokenizer-12Hz-Trainer/blob/main/src/trainer.py#L450) is what changes the setting to 48kHz (that's all).

> Is it possible to train it on music? (Just for fun experiment) 

The original model weights are likely designed for speech only, so I think fine-tuning would be difficult. While there might be a possibility with from-scratch training, you may need to include an encoder that isn't in my training code as a learning target as well.
